app/src/main/java/io/stormbird/wallet/service/KeyService.java

```diff
@@ -0,0 +1,1412 @@
+    {
+        //first recover the seed phrase from non-authlocked key. This removes the need to keep the seed phrase as a member on the heap
+        // - making the key operation more secure
+        String seedPhrase = unpackMnemonic(currentKeyAddress, UPGRADE_HD_KEY);
+        HDWallet newWallet = new HDWallet(seedPhrase, "");
+        storeHDKey(newWallet, Operation.IMPORT_HD_KEY);
+    }
+
+    private UpgradeKeyResult upgradeKey()
+    {
+        try
+        {
+            WalletType type = Wallet.getKeystoreType(context, currentKeyAddress);
+            //now check it's an unlocked key
+            String existingEncryptedKey = getEncryptedFilePath(currentKeyAddress);
+            if (existingEncryptedKey != null && (existingEncryptedKey.contains(NO_AUTH_LABEL) || type == KEYSTORE_LEGACY))
+            {
+                Operation keyOperation;
+                String secretData = null;
+
+                switch (type)
+                {
+                    case KEYSTORE:
+                        secretData = unpackMnemonic(currentKeyAddress, Operation.UPGRADE_HD_KEY);
+                        keyOperation = Operation.UPGRADE_KEYSTORE_KEY;
+                        break;
+                    case KEYSTORE_LEGACY:
+                        secretData = new String(getLegacyPassword(context, currentKeyAddress));
+                        keyOperation = Operation.UPGRADE_KEYSTORE_KEY;
+                        break;
+                    case HDKEY:
+                        secretData = unpackMnemonic(currentKeyAddress, Operation.UPGRADE_HD_KEY);
+                        keyOperation = Operation.UPGRADE_HD_KEY;
+                        break;
+                    default:
+                        return UpgradeKeyResult.ERROR;
+                }
+
+                if (secretData == null)
+                    return UpgradeKeyResult.ERROR;
+
+                storeEncryptedBytes(currentKeyAddress, secretData.getBytes(), keyOperation);
+
+                return UpgradeKeyResult.REQUESTING_SECURITY;
+            }
+            else
+            {
+                return UpgradeKeyResult.ALREADY_LOCKED;
+            }
+        }
+        catch (ServiceErrorException e)
+        {
+            //Legacy keystore error
+            if (!BuildConfig.DEBUG) Crashlytics.logException(e);
+            e.printStackTrace();
+            return UpgradeKeyResult.ERROR;
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+            return UpgradeKeyResult.ERROR;
+        }
+    }
+
+    private synchronized void storeHDKey(HDWallet newWallet, Operation operation)
+    {
+        PrivateKey pk = newWallet.getKeyForCoin(CoinType.ETHEREUM);
+        currentKeyAddress = CoinType.ETHEREUM.deriveAddress(pk);
+
+        storeEncryptedBytes(currentKeyAddress, newWallet.mnemonic().getBytes(), operation);
+    }
+
+    private synchronized void storeEncryptedBytes(String address, byte[] data, Operation operation)
```

Comment:
There are a lot of different flows that pass through here and it's difficult to review each possible one and ensure that, for example, keys and files are always deleted when necessary. I understand that this is done to limit code duplication, but IMO the code is too complex.